src/crimson/osd/recovery_backend.h

```diff
@@ -91,6 +103,10 @@ class RecoveryBackend {
     for (auto& [soid, recovery_waiter] : recovering) {
       recovery_waiter->stop();
     }
+    for (auto& [soid, promise] : unfound) {
+      promise.set_exception(
+	crimson::common::system_shutdown_exception());
+    }
```

Comment:
We should probably make sure that we drop the waiting requests in this case. I'm not sure this is handled, am I missing something?

Reply:
I think, by setting exceptions to the `unfound` promises, requests waiting on it will receive the exceptions, so the whole left client request processing will be skipped, which I think is equivalent to dropping the requests. Am I right?

Reply:
Yes, this is right. My concern is whether client request is actually handling exceptions correctly by dropping them. What do you think?

Reply:
Yes, this is right. My concern is whether client request actually handles exceptions correctly by dropping them. What do you think?

Reply:
Since the client request code is now using coroutines, IIUC, exceptions should be handled explicitly (aka by try-catch). See: https://docs.seastar.io/master/tutorial.html#exceptions-in-coroutines
(CC: @athanatos)

My suggestion is to change `add_unfound` to return ertr instead. That way, we could handle the error explicitly without try-catch.
  
```
    auto unfound = co_await do_recover_missing(
      pg, m->get_hobj().get_head(), m->get_reqid());
    if (unfound) {
      DEBUGDPP("{} unfound, hang it for now", *pg, m->get_hobj().get_head());
      co_await
        pg->get_recovery_backend()->add_unfound(m->get_hobj().get_head())
        ).handle_error_interruptible(... {
          DEBUG("dropping ...");
          co_return;
        });
    }
```

We can also catch the exceptions from the couroutine in order to drop the requests instead.

Reply:
will do.

Reply:
> Since the client request code is now using coroutines, IIUC, exceptions should be handled explicitly (aka by try-catch). See: https://docs.seastar.io/master/tutorial.html#exceptions-in-coroutines (CC: @athanatos)
> 
> My suggestion is to change `add_unfound` to return ertr instead. That way, we could handle the error explicitly without try-catch.
> 
> ```
>     auto unfound = co_await do_recover_missing(
>       pg, m->get_hobj().get_head(), m->get_reqid());
>     if (unfound) {
>       DEBUGDPP("{} unfound, hang it for now", *pg, m->get_hobj().get_head());
>       co_await
>         pg->get_recovery_backend()->add_unfound(m->get_hobj().get_head())
>         ).handle_error_interruptible(... {
>           DEBUG("dropping ...");
>           co_return;
>         });
>     }
> ```
> 
> We can also catch the exceptions from the couroutine in order to drop the requests instead.

@Matan-B After a second thought, the exception "system_shutdown_exception" we set to the `unfound` promises is actually an interrupt that will be handle by the `interruptor::with_interruption()` method. So, we are not leaking an exceptions.